### DISCUSSION TITLE
golang driver serialization API

### DISCUSSION THREAD

**Discussion by fmoor** [geldata/gel#1915] [General]
currently the golang driver serializes/de-serializes objects using [reflect](https://golang.org/pkg/reflect/) with a similar interface to  go's [json package](https://golang.org/pkg/encoding/json/#Unmarshal). Looking at other go database drivers (see list below) this seems to be the way it is done in go. 

- [mongo](https://pkg.go.dev/github.com/mongodb/mongo-go-driver@v1.2.1/mongo#Cursor.Decode)
- [mgo](https://pkg.go.dev/github.com/globalsign/mgo#Query.One)
- [pgx](https://pkg.go.dev/github.com/jackc/pgx#Rows.Scan)
- [pg](https://pkg.go.dev/github.com/go-pg/pg#Scan)
- [redigo](https://pkg.go.dev/github.com/gomodule/redigo/redis#Scan)

The only driver I found so far that has a significantly different API is [redis](https://pkg.go.dev/github.com/go-redis/redis#BoolCmd.Result) but I'm not sure this unique approach would be appropriate for EdgeDB since it would require statically defined interfaces for every possible return type.

If anyone has other ideas about how the serialization aspect of this driver might work I'd love to hear them!

# Example
note that the [struct tags](https://golang.org/pkg/reflect/#StructTag) are only required if the field name spelling in the schema is different than the struct. Because go only exports capitalized names struct fields that should be populated by the edgedb driver must be capitalized.

```go
type Person struct {
    FirstName string `edgedb:"first_name"`
    LastName  string `edgedb:"last_name"`
}

type Movie struct {
    Title    string   `edgedb:"title"`
    Year     int64    `edgedb:"year"`
    Director Person   `edgedb:"director"`
    Actors   []Person `edgedb:"actors"`
}

query := `
SELECT Movie {
        title,
        year,
        director: {
                first_name,
                last_name
        },
        actors: {
                first_name,
                last_name
        }
}
`

var results []Movie
client.Query(ctx, query, &results)

fmt.Println(results)
```

**Comment by 1st1** [comment]
I found this example for how to unmarshal JSON in go:

```go
func main() {
	var jsonBlob = []byte(`[
	{"Name": "Platypus", "Order": "Monotremata"},
	{"Name": "Quoll",    "Order": "Dasyuromorphia"}
]`)
	type Animal struct {
		Name  string
		Order string
	}
	var animals []Animal
	err := json.Unmarshal(jsonBlob, &animals)
	if err != nil {
		fmt.Println("error:", err)
	}
	fmt.Printf("%+v", animals)
}
```

Could you share a few examples on how that would work with EdgeQL queries and our API?

**Reply by fmoor** [reply]
Sure thing! I added an example.

**Comment by tailhook** [comment]
I think reflection is a way to go.

We also are probably going to have some codegen for query builder in TypeScript (and maybe Python). Although, in TypeScript we are going to generate structures, in Go we can probably generate deserialization code too (like in Rust, except in a different way :) ). That could be helpful for optimizing subset of very hot requests and other ones should be good with reflection.

**Reply by fmoor** [reply]
@tailhook are you thinking something similar to [ffjson](https://github.com/pquerna/ffjson) or [easyjson](https://github.com/mailru/easyjson)? Given a query generate source code with a struct(s) and serialization logic? or something along those lines?

**Reply by tailhook** [reply]
Yes

**Comment by fmoor** [comment]
I had been reading that reflection was slow. You can see just how much slower it is in these [benchmark results](https://github.com/edgedb/webapp-bench/pull/13#issuecomment-718282195). Compare `Edgedb Go Structs` to `Edgedb Go JSON`. The structs benchmark uses reflection where the json one doesn't.

**Reply by 1st1** [reply]
> The structs benchmark uses reflection where the json one doesn't.

This isn't about reflection alone. A bulk of performance is lost in binary protocol decoding and types unmarshaling, for all drivers. That said, we should optimize *how* we use the reflection.

Right now, you do this (I might be mistaken here, correct me if I'm wrong):

* You prepare the query and build codecs for it (stored in `Client.codecCache`)
* Every codec has a `Decode()` method
* You then execute the query, parsing the incoming binary data into Golang objects.
* Then you reflect the container type the user provided, and recursively traverse the parsed data guided by your reflection data, building new objects of user type and copying data into them.

This is indeed slow, not just because of reflection, but also of all the overhead associated with creating 2x objects (one for incoming data, one for actual returned results) & copying all of that.

How I'd recommend to do this instead:

* You prepare the query -- you have a full type descriptor now, more importantly you have the IN and OUT unique IDs for the data types.

* When you execute the query, the code knows the type of the receiving container. At this stage, ideally, we need to find a way to:

  * Get some unique identifier of the receiving container's type. **This is the part I'm not sure about -- is this possible in go?**
  * Combine that unique identifier with OUT codec's ID to form the cache key.
  * Lookup if there's a cached codec for that cache key. If not - build one.
  * We start actually using reflection when we build our codec -- we build the codecs tree with the knowledge of the receiving type's layout.

So when we start decoding the data we decode it right away into the struct layout the user expects. This way we avoid creating extra objects and using reflection to copy data.

Would it be possible to implement this?

**Reply by 1st1** [reply]
> Get some unique identifier of the receiving container's type. This is the part I'm not sure about -- is this possible in go?

Maybe there's pointer to its memory location -- we could use that as a key.

**Reply by fmoor** [reply]
:thinking: this seems possible. You can get the type of a value in go with [this](https://golang.org/pkg/reflect/#Value.Type).

**Reply by 1st1** [reply]
> this seems possible.

That's great, as it's the only way to get great performance.

**Reply by 1st1** [reply]
Also while at it, we should validate that the receiving struct actually matches what will come from the server, e.g.:

```go
type Person struct {
    FirstName uint8 `edgedb:"first_name"`  // should be string, not uint8
    LastName  string `edgedb:"last_name"`
}
```

I guess we should also require the query shape to map perfectly to the Go shape -- e.g. there shouldn't any unmapped properties/links and the scalar types must match.

**Reply by 1st1** [reply]
Maybe even the order of fields should match the order of links/properties.

**Reply by fmoor** [reply]
> Maybe even the order of fields should match the order of links/properties.

This seems like it would be surprising from a user perspective. Even requiring perfect mapping seems unusual.

**Reply by 1st1** [reply]
> This seems like it would be surprising from a user perspective. Even requiring perfect mapping seems unusual.

Wouldn't having matching order of fields make unpacking faster? Or the Go reflection API wouldn't care?

**Reply by fmoor** [reply]
It would make the codec building faster, but I don't think it would make a difference for decoding query results. We'll see how it shakes out in practice though.

**Reply by 1st1** [reply]
> Even requiring perfect mapping seems unusual.

Sure, we can relax this and only require that for all link/properties in an object shape there's a mapped field in the corresponding struct.

**Reply by 1st1** [reply]
> We'll see how it shakes out in practice though.

Yah, let's benchmark.

**Reply by fmoor** [reply]
:thinking: If we require an exact match between struct and query schema then there doesn't seem to be a way for the user to opt out of receiving `id` fields. Not sure this is important just noting.

**Reply by tailhook** [reply]
I think this is the case for making `id` optional too, along with `__tid__` and `__tname__` (https://github.com/edgedb/edgedb/issues/1350). But @elprans  proposed that all shapes  be "inherited" from some "base EdgeDB object" (which contains id).

**Reply by 1st1** [reply]
The problem is queries like `SELECT User` or `SELECT User { friends }`, which are perfectly valid and sensible in EdgeQL, but without an implicit ID they would result in empty objects on the client side. Which at least for dynamic programming languages like JS and Python doesn't make a whole lot of sense. So I suggest to simply ignore the "id" property. If ignoring it is too much hassle we can consider adding a knob to disable it (but it will be enabled by default).
